lightning/src/offers/flow.rs

```diff
@@ -1166,6 +1166,41 @@ where
 		Ok(())
 	}
 
+	/// If we are holding an HTLC on behalf of an often-offline sender, this method allows us to
+	/// create a path for the sender to use as the reply path when they send the recipient a
+	/// [`HeldHtlcAvailable`] onion message, so the recipient's [`ReleaseHeldHtlc`] response will be
+	/// received to our node.
+	///
+	/// [`ReleaseHeldHtlc`]: crate::onion_message::async_payments::ReleaseHeldHtlc
+	pub fn path_for_release_held_htlc<ES: Deref>(
+		&self, intercept_id: InterceptId, entropy: ES,
+	) -> BlindedMessagePath
+	where
+		ES::Target: EntropySource,
+	{
+		let peers = self.peers_cache.lock().unwrap().clone();
+		let context =
+			MessageContext::AsyncPayments(AsyncPaymentsContext::ReleaseHeldHtlc { intercept_id });
+		if let Ok(mut paths) = self.create_blinded_paths(peers, context.clone()) {
+			if let Some(path) = paths.pop() {
+				return path;
+			}
+		}
```

Comment:
So this actually introduces a lock order violation that was surfaced in the #4046's followup testing. It looks like it's because we're holding a peer state lock while this method is called, and the `DefaultMessageRouter` takes some locks internally to the `NetworkGraph` state. Looking into it, not sure yet if there's a nice solution.